lib/recursive-methods.rb

```diff
@@ -1,49 +1,98 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) => n operation will be needed! when n increases, time increases
+# Space complexity: O(n) => n number of recursive call
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
-end
 
-# Time complexity: ?
-# Space complexity: ?
-def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+    return 1 if n == 0
+    raise ArgumentError if n < 0
+
+    return n * factorial(n - 1)
 end
+#==========================================
+# Time complexity: O(n) => n operation will be needed!
+# Space complexity: O(n^2) => we have a slice method and with each recursive call a new string is created.
+def reverse(s)
+    return s if s.length <= 1
 
-# Time complexity: ?
-# Space complexity: ?
+    return s[-1] + reverse(s[0..-2])
+end
+# =========================================
+def reverse_inplace_helper(s, first = 0, last = s.length - 1)
+    if first < last
+        s[first], s[last] = s[last], s[first]
+        return reverse_inplace_helper(s, first + 1, last - 1)
+    else
+        return s
+    end
+end
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+    return s if s.length <= 1
+    reverse_inplace_helper(s)
 end
+#===================================
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+    return n if n == 0
+    return 2 + bunny(n - 1)
 end
-
-# Time complexity: ?
-# Space complexity: ?
+#======================================
+# Time complexity: O(n)
+# Space complexity: O(n^2) => slice method - each recursive call create a new string
 def nested(s)
```

Comment:
👍 , however the time/space complexity is O(n^2) for both.  Can you see a way, like reverse in place to do this with O(n)?